### PR TITLE
Update to xealousbrown.

### PR DESCRIPTION
**Pull request by alex-ong**
5-13ms Latency decrease, 4x scan rate improvement.
(CUSTOM_MATRIX = lite) is a really great feature!

Updated XealousBrown, various performance improvements.

## Description

* Decreased USB polling latency via setting rate from 10ms to 1ms
* Finally incorporated BOOTMAGIC so I can flash without opening the keyboard up.
* Used Eager_PK rather than sym_g which has 5ms lead lag time.
* Used CUSTOM_MATRIX = lite to increase polling rate using optimizations
* Changed pin polling code to pre-compiled scan for entire column.

## Types of Changes

- [ ] Core
- [ ] Bugfix
- [ ] New feature
- [x] Enhancement/optimization
- [x] Keyboard (addition or update)
- [ ] Keymap/layout/userspace (addition or update)
- [ ] Documentation

## Checklist

- [x] My code follows the code style of this project.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [ ] I have read the [**CONTRIBUTING** document](https://docs.qmk.fm/#/contributing).
- [ ] I have added tests to cover my changes.
- [x] I have tested the changes and verified that they work and don't break anything (as well as I can manage).

### CODE REVIEW

**Comment by zvecr** [comment]
While I like to see people using custom matrix lite, you might get the bulk of the performance gains from another solution.

Something like #8216, and setting the new define to 1. While this wont be the same as two noops, it will cut out a significant portion of the idle time without having to jump to custom code.

**Comment by alex-ong** [comment]
> While I like to see people using custom matrix lite, you might get the bulk of the performance gains from another solution.
> 
> Something like #8216, and setting the new define to 1. While this wont be the same as two noops, it will cut out a significant portion of the idle time without having to jump to custom code.

Thanks for the feedback! There are actually two parts to this; adding the nops "uncaps" the speed, allowing it to reach 10khz+. However the slow part is still the scanning.

The majority of the speedup is actually from get_cols(). I actually tested this over a year ago. The QMK way, though very modular, spends a lot of time setting rows and polling each column individually. It does quite a bit of lookup to achieve it. The TMK way you hardcode some arbitrary bitbash mask and it does it very quickly. 

I haven't tested it thoroughly on qmk recently, since I am just applying all my old hacks together. The net increase is from something like 1900 scanrate to 8000 scanrate.

It's nice that QMK has moved a lot of the hacks i made into core code (for example, setting usb polling rate to 1ms, CUSTOM_MATRIX = lite, Debouncing are all core features now)

**Comment by alex-ong** [comment]
Any easy way to generate such charts? I haven't tested all these speedups part by part recently so Ill try and document the speedups I did, I was going to write a blog post at some point anyway

**Comment by alex-ong** [comment]
So my optimizations are three things:
1) pin-settle time from wait_us(30) to wait_us(1) or nops or nothing at all
2) select_row() with a bit-bashy implementation
3) read_cols() with a bit-bashy implementation
4) unselect_rows() with a bit-bashy implementation.

First, overall performance compared to qmk matrix.c:
| pin-settle  | matrix_scan | scan_rate | Notes                                        |
|-------------|-------------|-----------|----------------------------------------------|
| wait_us(30) | qmk_default | 1900      | standard qmk implementation.                 |
| wait_us(1)  | qmk_default | 2635      | Changed wait_us by changing quantum/matrix.c |
| nop; nop;   | qmk_default | 2590      | Replaced wait_us in quantum/matrix.c         |
| (no settle) | qmk_default | 2670      | Deleted wait_us in quantum/matrix.c          |
| wait_us(30) | bit-bash    | 3662      | custom_matrix = lite                         |
| wait_us(1)  | bit-bash    | 7290      |                                              |
| nop; nop;   | bit-bash    | 8200      |                                              |
| (no settle) | bit-bash    | 8250      |                                              |

bit-bash means 2,3,4 use the "optimized" version.

Next table shows the performance difference of each part inside the custom_matrix implementation:
| pin_settle | select_row | read_cols | unselect_rows | scan_rate |
|------------|------------|-----------|---------------|-----------|
| nop; nop;  | fast       | fast      | fast          | 8250      |
| nop; nop;  | "qmk"      | fast      | "qmk"         | 5200      |
| nop; nop;  | "qmk"      | "qmk"     | "qmk"         | 2548      |

"qmk" represents an almost copy paste from qmk's matrix.c back into custom_matrix_lite, with slight modifications. you can see that the last option (2548) roughly matches qmk's real matrix implementation (2590).

But wait! There's more!
Changing a few compiler options also helps.
1) in TMK's rules.mk, you can enable Optimization level for speed vs size
2) you can enable `LINK_TIME_OPTIMIZATION_ENABLE = yes` for smaller/slightly faster binaries

| TMK's OPT   | LINK_TIME_OPTIMIZATION_ENABLE | scan_rate  | binary size |
|-------------|-------------------------------|------------|-------------|
| 3           | null (off)                    | 12036      | 21814       |
| 2           | null (off)                    | 11963      | 19488       |
| 1           | null (off)                    | 11964      | 20744       |
| 0 (off)     | null (off)                    | no_compile |  -          |
| s (default) | null (off)                    | 8212       | 17860       |
| 3           | yes                           | 13674      | 20634       |
| 2           | yes                           | 13655      | 17324       |
| 1           | yes                           | 9142       | 18952       |
| s (default) | yes                           | 8564       | 15926       |

Anyway its good that I sat down and did the research.

File: keyboards/handwired/xealousbrown/rules.mk [hunk 1]
```diff
@@ -14,17 +14,27 @@ BOOTLOADER = caterina
...
 #
-BOOTMAGIC_ENABLE = no      # Virtual DIP switch configuration(+1000)
-MOUSEKEY_ENABLE = no       # Mouse keys(+4700)
-EXTRAKEY_ENABLE = yes       # Audio control and System control(+450)
-CONSOLE_ENABLE = yes        # Console for debug(+400)
-COMMAND_ENABLE = yes        # Commands for debug and configuration
+BOOTMAGIC_ENABLE = full             # Virtual DIP switch configuration(+1000)
+MOUSEKEY_ENABLE = no                # Mouse keys(+4700)
+EXTRAKEY_ENABLE = yes               # Audio control and System control(+450)
+CONSOLE_ENABLE = yes                # Console for debug(+400)
```
Comment (noroadsleft) [inline keyboards/handwired/xealousbrown/rules.mk:20]:
```suggestion
BOOTMAGIC_ENABLE = full             # Virtual DIP switch configuration
MOUSEKEY_ENABLE = no                # Mouse keys
EXTRAKEY_ENABLE = yes               # Audio control and System control
CONSOLE_ENABLE = yes                # Console for debug
```

File: keyboards/handwired/xealousbrown/rules.mk [hunk 2]
```diff
@@ -14,17 +14,27 @@ BOOTLOADER = caterina
...
+UNICODE_ENABLE = no                 # Unicode
+BLUETOOTH_ENABLE = no               # Enable Bluetooth with the Adafruit EZ-Key HID
+AUDIO_ENABLE = no                   # Audio output on port C6
+SPACE_CADET_ENABLE = no             # Unneeded feature.
+
+
+# special sauce for this keyboard
+DEBOUNCE_TYPE = eager_pk            # Debounce using eager_pk.
+CUSTOM_MATRIX = lite                # Custom matrix that polls at 7000hz instead of a measly 2000hz.
+LINK_TIME_OPTIMIZATION_ENABLE = yes # smaller binary
```
Comment (noroadsleft) [inline keyboards/handwired/xealousbrown/rules.mk:37]: Totally optional, but you could this for the same result. (`LTO_ENABLE` is an alias in QMK for `LINK_TIME_OPTIMIZATION_ENABLE`.)

```suggestion
LTO_ENABLE = yes                    # smaller binary
```

**Comment by alex-ong** [comment]
Thanks for looking at it!

**Comment by noroadsleft** [comment]
Thanks!

As an aside, this board is a standard ANSI 60%, correct?

**Comment by alex-ong** [comment]
Yep, there's something about storing my bindings in user's/key-maps right?

**Comment by noroadsleft** [comment]
There is, but that wasn't what prompted the question.

I maintain the default keymaps over on QMK Configurator, and every time I pass by this one in my updates there, I'm reminded that your QMK Configurator implementation isn't correct – a user would have to assign the keys like [this](https://user-images.githubusercontent.com/18669334/80012737-9cd45f00-8482-11ea-9ca3-37e3d2cef7cd.png) to get the expected key assignments.

I'd like to fix it, but the correct way to do that per QMK's conventions requires that I edit your `keymap.c` and `xealousbrown.h` files, which I didn't want to do without permission.

As far as storing your bindings in `users/`, that's up to you, and whether that's worth your time really depends on how many QMK keyboards you own and how you use them. I currently only own one assembled QMK keyboard (have three more in various states of non-completion), so I don't use userspace.

**Comment by alex-ong** [comment]
@noroadsleft Thanks for that.

Feel free to correct it for me (aren't all devs lazy?); otherwise, `teach a man to fish...` is there a link to qmk conventions so i can correct it myself?

**Comment by alex-ong** [comment]
My guess is i have to change my keymap to this:
image

And then correct my pin mappings to suit?

**Comment by noroadsleft** [comment]
@alex-ong You posted while I was writing a response, but you've got it there. You just need that change to the `keymap.c`, and then to rearrange

https://github.com/qmk/qmk_firmware/blob/ccc8338b0e828f357a8051a729f2ccdbe3129fdb/keyboards/handwired/xealousbrown/xealousbrown.h#L6-L12

in the same way, so the keycodes that move are still assigned to the same arguments in the layout macro (Backspace as K4C, Backslash as K3C, etc.).

**Comment by alex-ong** [comment]
Alright i'll give it a shot tomorrow night, using GH60 layout as a guide to make sure those arrows in that diagram are right and i'm not missing any KC_NO's. Then will PR for someone to check and tell me i did it wrong lmao.

**Pull request merged**